File: backend/app/api/v1/schemas/environment.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.db.models.environment import EnvironmentStatus
from app.api.v1.schemas.system import SystemResponse


class EnvironmentCreate(BaseModel):
    name: str
    description: Optional[str] = None
    tier_id: int
    owner_user_id: int
    operations_group_id: Optional[int] = None
    expires_at: datetime
    status: EnvironmentStatus = EnvironmentStatus.ACTIVE
    custom_fields: Optional[dict] = None


class EnvironmentUpdate(BaseModel):
    # forbid: the six handover fields (access_url, connection_notes,
    # support_contact, sla_notes, known_limitations, decommission_notes) are
    # deliberately absent from this schema — PATCH /environments/{id}/handover
    # is their only write path. Without extra="forbid" Pydantic silently drops
    # unknown keys, which would let a client send them here too and have them
    # quietly ignored rather than rejected — a second, unguarded door.
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = None
    description: Optional[str] = None
    tier_id: Optional[int] = None
    owner_user_id: Optional[int] = None
    # Typed `Optional[int] = None` like every other field here — the
    # omitted-vs-null distinction is NOT carried by this type. It comes from
    # the service reading `model_fields_set`: an omitted key means "leave
    # alone", only an explicit null clears the group. Same contract as
    # expires_at.
    operations_group_id: Optional[int] = None
    expires_at: Optional[datetime] = None
    status: Optional[EnvironmentStatus] = None
    custom_fields: Optional[dict] = None


class EnvironmentHandoverUpdate(BaseModel):
    """The Welcome Pack's content. Six keys and no others.

    `extra="forbid"` is the safety property of this endpoint, not its
    authorization: whatever the permission rule says, a request body cannot
    reach tier_id, owner_user_id, operations_group_id or status through here.
    """

    model_config = ConfigDict(extra="forbid")

    access_url: Optional[str] = Field(default=None, max_length=500)
    connection_notes: Optional[str] = None
    support_contact: Optional[str] = Field(default=None, max_length=255)
    sla_notes: Optional[str] = None
    known_limitations: Optional[str] = None
    decommission_notes: Optional[str] = None


class EnvironmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: Optional[str] = None
    tier_id: int
    tier_name: str
    tier_color: Optional[str] = None
    owner_user_id: Optional[int] = None
    owner_username: Optional[str] = None
    operations_group_id: Optional[int] = None
    operations_group_name: Optional[str] = None
    expires_at: Optional[datetime] = None
    status: EnvironmentStatus
    reserved_now: bool = False
    # The Welcome Pack's content — read-only here; written only via
    # PATCH /environments/{id}/handover.
    access_url: Optional[str] = None
    connection_notes: Optional[str] = None
    support_contact: Optional[str] = None
    sla_notes: Optional[str] = None
    known_limitations: Optional[str] = None
    decommission_notes: Optional[str] = None
    tenant_id: int
    custom_fields: Optional[dict] = None
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_view(cls, view) -> "EnvironmentResponse":
        """Build from an environment_service.EnvironmentView.

        Display names travel with the row rather than being resolved by the
        browser against a separately-fetched collection — that collection is
        capped, so a `.find()` miss renders the entity as '—' and loses
        information no truncation banner can recover.
        """
        env = view.environment
        return cls(
            id=env.id,
            name=env.name,
            description=env.description,
            tier_id=env.tier_id,
            tier_name=view.tier_name,
            tier_color=view.tier_color,
            owner_user_id=env.owner_user_id,
            owner_username=view.owner_username,
            operations_group_id=env.operations_group_id,
            operations_group_name=view.operations_group_name,
            expires_at=env.expires_at,
            status=env.status,
            reserved_now=view.reserved_now,
            access_url=env.access_url,
            connection_notes=env.connection_notes,
            support_contact=env.support_contact,
            sla_notes=env.sla_notes,
            known_limitations=env.known_limitations,
            decommission_notes=env.decommission_notes,
            tenant_id=env.tenant_id,
            custom_fields=env.custom_fields,
            created_at=env.created_at,
            updated_at=env.updated_at,
        )


class EnvironmentSystemCreate(BaseModel):
    system_id: int


class EnvironmentSystemUpdate(BaseModel):
    pass  # reserved for future fields


class EnvironmentSystemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    environment_id: int
    system_id: int
    system: SystemResponse


class SystemSummary(BaseModel):
    id: int
    name: str
    description: Optional[str] = None


class EnvironmentSystemsResponse(BaseModel):
    """Response for GET /environments/{env_id}/systems — includes missing systems."""
    systems: list[EnvironmentSystemResponse]
    missing_systems: list[SystemSummary]


class VersionSummary(BaseModel):
    build_identifier: str
    version_label: str
    installed_at: datetime


class EnvironmentSubsystemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    environment_id: int
    subsystem_id: int
    subsystem_name: str
    component_type: str
    component_type_definition_id: Optional[int] = None
    component_type_definition_name: Optional[str] = None
    technology: Optional[str] = None
    system_id: int
    system_name: str
    is_mocked: bool
    mock_notes: Optional[str] = None
    custom_fields: Optional[dict] = None
    latest_version: Optional[VersionSummary] = None


class EnvironmentSubsystemUpdate(BaseModel):
    is_mocked: Optional[bool] = None
    mock_notes: Optional[str] = None
    component_type_definition_id: Optional[int] = None
    custom_fields: Optional[dict] = None


class EnvSubsystemHostRef(BaseModel):
    """A host an environment subsystem is deployed on, for the topology diagram."""
    infrastructure_component_id: int
    name: str
    component_type: str          # InfrastructureComponentType value (str-enum)
    role: Optional[str] = None


class EnvSubsystemNode(BaseModel):
    """Subsystem node for the environment topology response."""
    id: int
    name: str
    component_type: str
    technology: Optional[str] = None
    system_id: int
    is_mocked: bool
    hosts: list[EnvSubsystemHostRef] = []


from app.api.v1.schemas.dependency import ComponentDependencyResponse  # noqa: E402


class EnvironmentTopologyResponse(BaseModel):
    environment_id: int
    subsystems: list[EnvSubsystemNode]
    dependencies: list[ComponentDependencyResponse]
    system_names: dict[str, str]
    outside_subsystems: list[EnvSubsystemNode]
    outside_dependencies: list[ComponentDependencyResponse]

File: frontend/src/types/environment.ts
import type { DecommissionState } from './decommission';

export type EnvironmentStatus = 'active' | 'inactive' | 'maintenance' | 'decommissioned';

export interface EnvironmentResponse {
  id: number;
  name: string;
  description: string | null;
  tier_id: number;
  tier_name: string;
  tier_color: string | null;
  owner_user_id: number | null;
  /**
   * The owner's display name, travelling with the row rather than being
   * resolved in the browser against a separately-fetched (and capped) user
   * collection — see docs/pagination.md.
   */
  owner_username: string | null;
  /** Null means "no expiry planned" — a legitimate state, not a missing value. */
  expires_at: string | null;
  reserved_now: boolean;
  /** B5 — no deployment or booking for the threshold. Advisory only. */
  idle: boolean;
  /** B5 — the live decommission's computed state, or null if there is none. */
  decommission_state: DecommissionState | null;
  status: EnvironmentStatus;
  tenant_id: number;
  custom_fields: Record<string, unknown> | null;
  operations_group_id: number | null;
  operations_group_name: string | null;
  /** The Welcome Pack's content — read-only here; written only via PATCH /environments/{id}/handover. */
  access_url: string | null;
  connection_notes: string | null;
  support_contact: string | null;
  sla_notes: string | null;
  known_limitations: string | null;
  decommission_notes: string | null;
  /**
   * B2. NULL means "no pattern applies" — it counts as COMPLIANT, never as
   * unknown and never as failing. No policy row, a disabled policy, or a
   * policy with no `name_pattern` all read as null.
   */
  name_compliant?: boolean | null;
  /** Derived on read from the stored verdict, `created_at` and the policy. Advisory: it blocks nothing. */
  quarantined?: boolean;
  compliance_gaps?: string[];
  created_at: string;
  updated_at: string;
}

export interface EnvironmentNamingPolicy {
  is_enabled: boolean;
  name_pattern: string | null;
  name_pattern_example: string | null;
  required_attributes: string[];
  grace_days: number;
  effective_from: string;
}

/**
 * What PUT accepts, which is NOT what GET returns: `effective_from` is the
 * server's to set (it bumps whenever the pattern or the attribute list changes,
 * in either direction), and `EnvironmentNamingPolicyUpdate` declares
 * `extra="forbid"` — so echoing the whole read model back is a 422, not a
 * harmless extra key.
 */
export type EnvironmentNamingPolicyUpdate = Omit<EnvironmentNamingPolicy, 'effective_from'>;

export interface EnvironmentNamingPolicyPreview {
  total_environments: number;
  in_gap: number;
  quarantined_now: number;
  /**
   * Capped server-side at 20. The COUNTS above are exact; this is a sample, and
   * anything rendering it must say so rather than imply it is the whole set.
   */
  sample_names: string[];
}

export interface SystemSummary {
  id: number;
  name: string;
  description: string | null;
}

export interface EnvironmentSystemResponse {
  id: number;
  environment_id: number;
  system_id: number;
  system: {
    id: number;
    name: string;
    description: string | null;
    github_repository_url: string | null;
  };
}

export interface EnvironmentSystemsResponse {
  systems: EnvironmentSystemResponse[];
  missing_systems: SystemSummary[];
}

export interface EnvironmentCreate {
  name: string;
  description?: string;
  tier_id: number;
  owner_user_id: number;
  /** Required on create; `PATCH` may clear it later (see EnvironmentUpdate). */
  expires_at: string;
  status?: EnvironmentStatus;
  custom_fields?: Record<string, unknown>;
  operations_group_id?: number | null;
}

export interface EnvironmentUpdate {
  name?: string;
  description?: string;
  tier_id?: number;
  owner_user_id?: number;
  /**
   * `null` clears the expiry — "no expiry planned", which the backend treats
   * as a legitimate state and deliberately excludes from its compliance rule.
   * Omit the key to leave the stored value untouched.
   */
  expires_at?: string | null;
  status?: EnvironmentStatus;
  custom_fields?: Record<string, unknown>;
  /**
   * `null` clears the group. The backend keys on `model_fields_set`, so an
   * omitted key means "leave alone" — only an explicit null clears it.
   */
  operations_group_id?: number | null;
}

export interface EnvironmentSystemCreate {
  system_id: number;
}

export interface EnvironmentSystemUpdate {
  // reserved for future fields
}

export interface VersionSummary {
  build_id: string;
  version_label: string;
  installed_at: string;
}

export interface EnvironmentSubsystemResponse {
  id: number;
  environment_id: number;
  subsystem_id: number;
  subsystem_name: string;
  component_type: string;
  component_type_definition_id: number | null;
  component_type_definition_name: string | null;
  technology: string | null;
  system_id: number;
  system_name: string;
  is_mocked: boolean;
  mock_notes: string | null;
  custom_fields: Record<string, unknown> | null;
  latest_version: VersionSummary | null;
}

export interface EnvironmentSubsystemUpdate {
  is_mocked?: boolean;
  mock_notes?: string | null;
  component_type_definition_id?: number | null;
  custom_fields?: Record<string, unknown> | null;
}

export interface EnvSubsystemHostRef {
  infrastructure_component_id: number;
  name: string;
  component_type: string;
  role: string | null;
}

export interface EnvSubsystemNode {
  id: number;
  name: string;
  component_type: string;
  technology: string | null;
  system_id: number;
  is_mocked: boolean;
  hosts: EnvSubsystemHostRef[]; // [] for outside subsystems
}

import type { ComponentDependencyResponse } from './dependency';

export interface EnvironmentTopologyData {
  environment_id: number;
  subsystems: EnvSubsystemNode[];
  dependencies: ComponentDependencyResponse[];
  system_names: Record<string, string>;
  outside_subsystems: EnvSubsystemNode[];
  outside_dependencies: ComponentDependencyResponse[];
}

// Alias for use in components (e.g., EnvironmentPicker)
export type Environment = EnvironmentResponse;
